File: backend/src/main/java/com/workflow/api/controller/SystemInternalController.java
package com.workflow.api.controller;

import com.workflow.api.dto.BizDataPageVO;
import com.workflow.api.dto.BizDataQueryRequest;
import com.workflow.api.dto.BizDataVO;
import com.workflow.api.dto.DataSourceMetadata;
import com.workflow.common.domain.PageResult;
import com.workflow.common.domain.R;
import com.workflow.common.exception.BusinessException;
import com.workflow.engine.datasource.BuiltInSystemSourceQueryService;
import com.workflow.engine.form.column.ColumnConfig;
import com.workflow.system.domain.dto.OrganizationCreateRequest;
import com.workflow.system.domain.dto.UserCreateRequest;
import com.workflow.system.domain.dto.UserQueryRequest;
import com.workflow.system.domain.vo.TreeNode;
import com.workflow.system.domain.vo.UserVO;
import com.workflow.system.service.OrganizationService;
import com.workflow.system.service.UserService;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * SYSTEM 内部 REST 控制器。
 * 部门树扁平化（parentId=root 时为空串）；用户分页；元数据只读标记；CRUD 委托。
 *
 * 这些端点不是给前端直接调的，而是<b>数据源 SPI 的另一半</b>：
 * {@code /api/v1/data-sources} 里类型为 SYSTEM 的数据源（部门树、用户、内建系统数据源）转发到这里。
 * 响应用 {@code BizDataVO} / {@code BizDataPageVO}，与业务表单数据同构，
 * 页面设计器才能用同一套渲染逻辑。
 */
@RestController
@RequestMapping("/api/v1/internal")
public class SystemInternalController {

    private final OrganizationService organizationService;
    private final UserService userService;
    private final BuiltInSystemSourceQueryService builtInSourceQuery;

    public SystemInternalController(OrganizationService organizationService,
                                    UserService userService,
                                    BuiltInSystemSourceQueryService builtInSourceQuery) {
        this.organizationService = organizationService;
        this.userService = userService;
        this.builtInSourceQuery = builtInSourceQuery;
    }

    // ==================== READ ====================

    /**
     * 部门树扁平化。根节点 parentId 为空串。
     * keyword 过滤 label 和 code（忽略大小写包含匹配）。
     */
    @GetMapping("/system/dept-tree")
    public R<BizDataPageVO> deptTree(@RequestParam(required = false) String keyword) {
        List<TreeNode> tree = organizationService.tree();
        List<BizDataVO> flattened = flattenTree(tree, keyword);
        BizDataPageVO page = new BizDataPageVO(flattened, flattened.size(), 0, flattened.size());
        return R.ok(page);
    }

    /**
     * 用户分页。
     */
    @GetMapping("/system/users")
    public R<BizDataPageVO> users(@RequestParam(required = false) String keyword,
                                  @RequestParam(defaultValue = "1") Integer page,
                                  @RequestParam(defaultValue = "20") Integer size) {
        int userServicePage = Math.max(page, 1);
        UserQueryRequest query = new UserQueryRequest(keyword, null, null, null, null, null, userServicePage, size);
        PageResult<UserVO> result = userService.list(query);
        List<BizDataVO> records = result.getRows().stream()
                .map(this::toUserRow)
                .collect(Collectors.toList());
        BizDataPageVO pageVo = new BizDataPageVO(records, result.getTotal(), userServicePage, size);
        return R.ok(pageVo);
    }

    /**
     * 根据 ID 查询用户。
     */
    @GetMapping("/system/users/{id}")
    public R<BizDataVO> getUser(@PathVariable String id) {
        UserVO user = userService.getById(Long.valueOf(id));
        if (user == null) {
            throw new BusinessException("用户不存在");
        }
        return R.ok(toUserRow(user));
    }

    // ==================== metadata ====================

    @GetMapping("/system/dept-tree/metadata")
    public R<DataSourceMetadata> deptTreeMetadata() {
        List<ColumnConfig> columns = List.of(
                columnConfig("id", "ID"),
                columnConfig("parentId", "父节点"),
                columnConfig("label", "名称"),
                columnConfig("code", "编码")
        );
        return R.ok(new DataSourceMetadata(columns, true));
    }

    @GetMapping("/system/users/metadata")
    public R<DataSourceMetadata> usersMetadata() {
        List<ColumnConfig> columns = List.of(
                columnConfig("id", "ID"),
                columnConfig("username", "用户名"),
                columnConfig("nickname", "昵称"),
                columnConfig("orgId", "组织ID"),
                columnConfig("orgName", "组织名称"),
                columnConfig("status", "状态")
        );
        return R.ok(new DataSourceMetadata(columns, true));
    }

    // ==================== CRUD: dept ====================

    @PostMapping("/system/dept")
    public R<BizDataVO> createDept(@RequestBody Map<String, Object> data) {
        String orgName = (String) data.get("orgName");
        String orgCode = (String) data.get("orgCode");
        OrganizationCreateRequest request = new OrganizationCreateRequest(null, orgName, orgCode, null, null);
        TreeNode node = organizationService.create(request);
        return R.ok(toDeptRow(node));
    }

    @DeleteMapping("/system/dept/{id}")
    public R<Void> deleteDept(@PathVariable String id) {
        organizationService.delete(Long.valueOf(id));
        return R.ok();
    }

    // ==================== CRUD: users ====================

    @PostMapping("/system/user")
    public R<BizDataVO> createUser(@RequestBody Map<String, Object> data) {
        String username = (String) data.get("username");
        String nickname = (String) data.get("nickname");
        Object orgIdObj = data.get("orgId");
        Long orgId = orgIdObj instanceof Number n ? n.longValue() : null;
        UserCreateRequest request = new UserCreateRequest(username, nickname, null, null, orgId, null, null);
        UserVO user = userService.create(request);
        return R.ok(toUserRow(user));
    }

    @DeleteMapping("/system/user/{id}")
    public R<Void> deleteUser(@PathVariable String id) {
        userService.delete(Long.valueOf(id));
        return R.ok();
    }

    // ==================== 内建系统数据源 REST 化取数（新 6 个，50-a） ====================
    //
    // 这些端点与数据源 SPI（adapter 直调服务）共享同一份实现（BuiltInSystemSourceQueryService），
    // 是 V39 迁移脚本预置的 params.list.action 的落点；也给前端/页面直连提供 REST 形状。取数均只读。

    /** 系统菜单列表（全量扁平化）。 */
    @GetMapping("/system/menus")
    public R<BizDataPageVO> systemMenus(@RequestParam(defaultValue = "1") Integer page,
                                        @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("sys-menus", listRequest(page, size)));
    }

    /** 系统菜单元数据。 */
    @GetMapping("/system/menus/metadata")
    public R<DataSourceMetadata> systemMenusMetadata() {
        return R.ok(sourceMetadata("sys-menus"));
    }

    /** 系统角色列表（标准分页）。 */
    @GetMapping("/system/roles")
    public R<BizDataPageVO> systemRoles(@RequestParam(defaultValue = "1") Integer page,
                                        @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("sys-roles", listRequest(page, size)));
    }

    /** 系统角色元数据。 */
    @GetMapping("/system/roles/metadata")
    public R<DataSourceMetadata> systemRolesMetadata() {
        return R.ok(sourceMetadata("sys-roles"));
    }

    /** 系统字典（类型）列表（标准分页）。 */
    @GetMapping("/system/dicts")
    public R<BizDataPageVO> systemDicts(@RequestParam(defaultValue = "1") Integer page,
                                        @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("sys-dicts", listRequest(page, size)));
    }

    /** 系统字典元数据。 */
    @GetMapping("/system/dicts/metadata")
    public R<DataSourceMetadata> systemDictsMetadata() {
        return R.ok(sourceMetadata("sys-dicts"));
    }

    /** 流程定义列表（全量）。 */
    @GetMapping("/system/process/definitions")
    public R<BizDataPageVO> processDefinitions(@RequestParam(defaultValue = "1") Integer page,
                                               @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("process-definitions", listRequest(page, size)));
    }

    /** 流程定义元数据。 */
    @GetMapping("/system/process/definitions/metadata")
    public R<DataSourceMetadata> processDefinitionsMetadata() {
        return R.ok(sourceMetadata("process-definitions"));
    }

    /** 流程实例列表（标准分页，运行中）。 */
    @GetMapping("/system/process/instances")
    public R<BizDataPageVO> processInstances(@RequestParam(defaultValue = "1") Integer page,
                                             @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("process-instances", listRequest(page, size)));
    }

    /** 流程实例元数据。 */
    @GetMapping("/system/process/instances/metadata")
    public R<DataSourceMetadata> processInstancesMetadata() {
        return R.ok(sourceMetadata("process-instances"));
    }

    /** 待办任务列表（标准分页，按当前登录人过滤）。 */
    @GetMapping("/system/process/todo-tasks")
    public R<BizDataPageVO> processTodoTasks(@RequestParam(defaultValue = "1") Integer page,
                                             @RequestParam(defaultValue = "20") Integer size) {
        return R.ok(builtInSourceQuery.query("todo-tasks", listRequest(page, size)));
    }

    /** 待办任务元数据。 */
    @GetMapping("/system/process/todo-tasks/metadata")
    public R<DataSourceMetadata> processTodoTasksMetadata() {
        return R.ok(sourceMetadata("todo-tasks"));
    }

    // ==================== helpers ====================

    private ColumnConfig columnConfig(String key, String label) {
        ColumnConfig c = new ColumnConfig();
        c.setKey(key);
        c.setLabel(label);
        return c;
    }

    /** 内建数据源取数请求（page/size 缺省与 users 端点同默认：1/20）。 */
    private BizDataQueryRequest listRequest(Integer page, Integer size) {
        BizDataQueryRequest req = new BizDataQueryRequest();
        req.setPage(Math.max(page == null ? 1 : page, 1));
        req.setSize(size == null ? 20 : size);
        return req;
    }

    /**
     * 内建数据源元数据（列来自目录常量 BuiltInSystemSources）。
     * 与既有 deptTreeMetadata/usersMetadata 同构传 writable=true。
     */
    private DataSourceMetadata sourceMetadata(String sourceKey) {
        List<ColumnConfig> columns = new ArrayList<>();
        for (ColumnConfig c : builtInSourceQuery.columnsOf(sourceKey)) {
            columns.add(columnConfig(c.getKey(), c.getLabel()));
        }
        return new DataSourceMetadata(columns, true);
    }

    private List<BizDataVO> flattenTree(List<TreeNode> nodes, String keyword) {
        List<BizDataVO> result = new ArrayList<>();
        if (nodes == null) return result;
        for (TreeNode node : nodes) {
            flattenNode(node, result, keyword);
        }
        return result;
    }

    private void flattenNode(TreeNode node, List<BizDataVO> result, String keyword) {
        if (node == null) return;
        boolean matches = keyword == null || keyword.isEmpty()
                || (node.label() != null && node.label().toLowerCase().contains(keyword.toLowerCase()))
                || (node.code() != null && node.code().toLowerCase().contains(keyword.toLowerCase()));
        if (matches) {
            result.add(toDeptRow(node));
        }
        if (node.children() != null) {
            for (TreeNode child : node.children()) {
                flattenNode(child, result, keyword);
            }
        }
    }

    private BizDataVO toDeptRow(TreeNode node) {
        Map<String, Object> data = new HashMap<>();
        data.put("parentId", node.parentId() != null ? String.valueOf(node.parentId()) : "");
        data.put("label", node.label());
        data.put("code", node.code());
        return new BizDataVO(String.valueOf(node.id()), data, null, null, null);
    }

    private BizDataVO toUserRow(UserVO user) {
        Map<String, Object> data = new HashMap<>();
        data.put("username", user.username());
        data.put("nickname", user.nickname());
        data.put("orgId", user.orgId());
        data.put("orgName", user.orgName());
        data.put("status", user.status());
        return new BizDataVO(String.valueOf(user.id()), data, null, null, null);
    }
}

File: backend-node/src/api/controllers/system-internal.controller.ts
import { Body, Controller, Delete, Get, Param, Post, Query } from '@nestjs/common'
import { R } from '../../common/domain/r'
import { BusinessException } from '../../common/exception/business-exception'
import {
  bizDataPageVO,
  bizDataVO,
  type BizDataPageVO,
  type BizDataVO,
} from '../../common/domain/biz-data'
import {
  columnConfig,
  dataSourceMetadata,
  type DataSourceMetadata,
} from '../../common/domain/column-config'
import { JavaStatusOk } from '../../framework/http/java-status.decorator'
import { bindIntProperty, integerQueryParam } from '../../framework/http/query-params'
import { SystemService, type TreeNodeVO, type UserVO } from '../../system/service/system.service'
import { SystemSourceQueryService } from '../../engine/datasource/service/system-source-query.service'
import { BUILT_IN_SOURCE_KEYS } from '../../engine/datasource/service/system-source-catalog'

/**
 * SYSTEM 内部 REST 控制器（对齐 Java `SystemInternalController`，前缀 `/api/v1/internal`）。
 *
 * 这些端点不是给前端直接调的，而是**数据源 SPI 的另一半**：
 * `/api/v1/data-sources` 里类型为 SYSTEM 的数据源（部门树、用户）转发到这里。
 * 所以它们的响应形状用 `BizDataVO` / `BizDataPageVO`，与业务表单数据同构 ——
 * 页面设计器才能用同一套渲染逻辑。
 *
 * 放在 api 层是因为 Java 侧就在 `com.workflow.api.controller`，
 * 而且它同时用到了 system 模块的服务；engine 层不得依赖 api。
 */
@Controller('api/v1/internal')
@JavaStatusOk()
export class SystemInternalController {
  constructor(
    private readonly userService: SystemService,
    private readonly systemSourceQuery: SystemSourceQueryService,
  ) {}

  /**
   * 部门树扁平化（根节点 `parentId` 为空串）。
   * `keyword` 忽略大小写匹配 label 与 code。
   *
   * ⚠️ 分页外壳的 `page`/`size` 都是**扁平化后的条数**，不是请求参数
   *    （Java 传的是 `new BizDataPageVO(flattened, flattened.size(), 0, flattened.size())`）。
   */
  @Get('system/dept-tree')
  async deptTree(@Query('keyword') keyword?: string): Promise<R<BizDataPageVO>> {
    const nodes = await this.userService.orgTree()
    const flattened = flattenTree(nodes, normalizeKeyword(keyword))
    return R.ok(bizDataPageVO(flattened, flattened.length, 0, flattened.length))
  }

  /** 用户分页。 */
  @Get('system/users')
  async users(
    @Query('keyword') keyword?: string,
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    const result = await this.userService.listUsersByUsername(
      normalizeKeyword(keyword),
      // Java 这里声明的是 `Integer`（不是 `int`）⇒ 失败消息里的类型名是 `java.lang.Integer`
      integerQueryParam(page, 'page', 1),
      integerQueryParam(size, 'size', 20),
    )
    return R.ok(
      bizDataPageVO(result.rows.map(toUserRow), result.total, result.page, result.size),
    )
  }

  // ⚠️ 路由声明顺序有语义：Nest 按声明顺序匹配，`system/users/metadata`
  //    必须排在 `system/users/:id` 之前，否则 `:id` 会吃掉 `metadata`，
  //    走到 `Number('metadata') === NaN` 拼出一条 `WHERE id = NaN` 的 SQL
  //    （契约比对抓到的真实故障，不是理论风险）。

  /** 部门树元数据（固定 4 列，不可写标记为 true）。 */
  @Get('system/dept-tree/metadata')
  deptTreeMetadata(): R<DataSourceMetadata> {
    return R.ok(
      dataSourceMetadata(
        [
          columnConfig('id', 'ID'),
          columnConfig('parentId', '父节点'),
          columnConfig('label', '名称'),
          columnConfig('code', '编码'),
        ],
        true,
      ),
    )
  }

  /** 用户元数据（固定 6 列）。 */
  @Get('system/users/metadata')
  usersMetadata(): R<DataSourceMetadata> {
    return R.ok(
      dataSourceMetadata(
        [
          columnConfig('id', 'ID'),
          columnConfig('username', '用户名'),
          columnConfig('nickname', '昵称'),
          columnConfig('orgId', '组织ID'),
          columnConfig('orgName', '组织名称'),
          columnConfig('status', '状态'),
        ],
        true,
      ),
    )
  }

  /** 按 id 查用户。 */
  @Get('system/users/:id')
  async getUser(@Param('id') id: string): Promise<R<BizDataVO>> {
    return R.ok(toUserRow(await this.userService.getUserById(Number(id))))
  }

  // ==================== 写（委托 system 模块，同样对齐 Java 的委托） ====================

  /**
   * 新建部门。
   *
   * ⚠️ Java 传的是 `new OrganizationCreateRequest(null, orgName, orgCode, null, null)`
   *    —— **parentId 恒为 null**、sortOrder/status 也不传（走服务端默认 0/1）。
   *    所以这个端点**只能建根节点**。请求体里给了 parentId 也不会生效。
   */
  @Post('system/dept')
  async createDept(@Body() body: Record<string, unknown> | null): Promise<R<BizDataVO>> {
    const node = await this.userService.createOrg({
      parentId: null,
      orgName: String(body?.orgName ?? ''),
      orgCode: String(body?.orgCode ?? ''),
      sortOrder: null,
      status: null,
    })
    return R.ok(toDeptRow(node))
  }

  /**
   * 删除部门（**软删除**）。
   *
   * ⚠️ 三重校验决定报哪条错误：不存在 → 存在子节点 → 该机构下存在用户。
   *    后两条都是业务异常（HTTP 200 + code 500）。
   */
  @Delete('system/dept/:id')
  async deleteDept(@Param('id') id: string): Promise<R<null>> {
    await this.userService.deleteOrg(Number(id))
    return R.ok()
  }

  /**
   * 新建用户。
   *
   * ⚠️ Java 传 `new UserCreateRequest(username, nickname, null, null, orgId, null, null)`
   *    —— email/phone/roleIds 恒为 null、status 走默认 1。
   *    `orgId` 只有是 Number 时才取，其余（含字符串）一律为 null。
   */
  @Post('system/user')
  async createUser(@Body() body: Record<string, unknown> | null): Promise<R<BizDataVO>> {
    const orgIdRaw = body?.orgId
    const user = await this.userService.createUser({
      username: String(body?.username ?? ''),
      nickname: String(body?.nickname ?? ''),
      email: null,
      phone: null,
      orgId: typeof orgIdRaw === 'number' ? orgIdRaw : null,
      roleIds: null,
      status: null,
    })
    return R.ok(toUserRow(user))
  }

  /** 删除用户（**软删除**）。 */
  @Delete('system/user/:id')
  async deleteUser(@Param('id') id: string): Promise<R<null>> {
    await this.userService.deleteUser(Number(id))
    return R.ok()
  }

  // ==================== 内建系统数据源 REST 化取数（新 6 个） ====================
  //
  // 这些端点与数据源 SPI（adapter 直调服务）共享同一份实现
  // （SystemSourceQueryService），是 `generateParams` 生成的 `params.list.action`
  // 的落点；也给前端/页面直连提供 REST 形状。取数均**只读**。
  //
  // ⚠️ 路由声明顺序有语义：`*/metadata` 与字面量段（process/...）必须排在
  // 对应参数段（*/:id）之前（Nest 按声明顺序匹配，同 users 的历史教训）。

  /** 系统菜单列表（全量扁平化）。 */
  @Get('system/menus')
  async systemMenus(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('sys-menus', page, size))
  }

  /** 系统菜单元数据。 */
  @Get('system/menus/metadata')
  async systemMenusMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('sys-menus'))
  }

  /** 系统菜单单条。 */
  @Get('system/menus/:id')
  async systemMenuById(@Param('id') id: string): Promise<R<BizDataVO>> {
    return R.ok(await this.sourceGet('sys-menus', id))
  }

  /** 系统角色列表（标准分页）。 */
  @Get('system/roles')
  async systemRoles(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('sys-roles', page, size))
  }

  /** 系统角色元数据。 */
  @Get('system/roles/metadata')
  async systemRolesMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('sys-roles'))
  }

  /** 系统角色单条。 */
  @Get('system/roles/:id')
  async systemRoleById(@Param('id') id: string): Promise<R<BizDataVO>> {
    return R.ok(await this.sourceGet('sys-roles', id))
  }

  /** 系统字典（类型）列表（标准分页）。 */
  @Get('system/dicts')
  async systemDicts(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('sys-dicts', page, size))
  }

  /** 系统字典元数据。 */
  @Get('system/dicts/metadata')
  async systemDictsMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('sys-dicts'))
  }

  /** 系统字典单条。 */
  @Get('system/dicts/:id')
  async systemDictById(@Param('id') id: string): Promise<R<BizDataVO>> {
    return R.ok(await this.sourceGet('sys-dicts', id))
  }

  /** 流程定义列表（全量）。 */
  @Get('system/process/definitions')
  async processDefinitions(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('process-definitions', page, size))
  }

  /** 流程定义元数据。 */
  @Get('system/process/definitions/metadata')
  async processDefinitionsMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('process-definitions'))
  }

  /** 流程实例列表（标准分页，运行中）。 */
  @Get('system/process/instances')
  async processInstances(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('process-instances', page, size))
  }

  /** 流程实例元数据。 */
  @Get('system/process/instances/metadata')
  async processInstancesMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('process-instances'))
  }

  /** 待办任务列表（标准分页，按当前登录人过滤）。 */
  @Get('system/process/todo-tasks')
  async processTodoTasks(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<BizDataPageVO>> {
    return R.ok(await this.sourceList('todo-tasks', page, size))
  }

  /** 待办任务元数据。 */
  @Get('system/process/todo-tasks/metadata')
  async processTodoTasksMetadata(): Promise<R<DataSourceMetadata>> {
    return R.ok(await this.sourceMetadata('todo-tasks'))
  }

  /** 内建数据源取数（page/size 缺省与 queryData 同默认：1/20）。 */
  private async sourceList(
    sourceKey: string,
    page?: string,
    size?: string,
  ): Promise<BizDataPageVO> {
    return this.systemSourceQuery.query(sourceKey, {
      filter: null,
      keyword: null,
      keywordColumn: null,
      sort: null,
      order: null,
      params: null,
      page: bindIntProperty(page, 'page', 1),
      size: bindIntProperty(size, 'size', 20),
    })
  }

  /**
   * 内建数据源元数据（列来自目录常量）。
   *
   * ⚠️ 第二参与既有 deptTreeMetadata/usersMetadata 同构传 `true`
   * （历史端点的既成行为）；数据管理页实际走 SPI 路径
   * `/data-sources/:id/metadata`（adapter 返回 writable=false 只读），
   * 这些 REST metadata 端点当前没有前端消费方，一致性优先。
   */
  private async sourceMetadata(sourceKey: string): Promise<DataSourceMetadata> {
    return dataSourceMetadata(
      (await this.systemSourceQuery.columnsOf(sourceKey)).map((column) =>
        columnConfig(String(column.key), String(column.label)),
      ),
      true,
    )
  }

  /** 内建数据源单条：全量取回后按 id 线性查找（对齐 adapter SYSTEM get 的语义）。 */
  private async sourceGet(sourceKey: string, id: string): Promise<BizDataVO> {
    if (!BUILT_IN_SOURCE_KEYS.has(sourceKey)) {
      throw new BusinessException(400, `未注册的系统数据源: ${sourceKey}`)
    }
    const all = await this.systemSourceQuery.query(sourceKey, {
      filter: null,
      keyword: null,
      keywordColumn: null,
      sort: null,
      order: null,
      params: null,
      page: 1,
      // 全量取回：size 给安全上限，full/paged 语义差异由服务内部处理
      size: 100000,
    })
    const row = all.records.find((item) => item.id === id)
    if (row === undefined) {
      throw new BusinessException(404, `系统数据不存在: ${id}`)
    }
    return row
  }
}

/** Java `keyword == null || keyword.isEmpty()` —— 注意**空白串不算空**，与 isBlank 不同。 */
function normalizeKeyword(keyword: string | undefined): string | null {
  if (keyword === undefined || keyword === '') return null
  return keyword
}

/**
 * 部门树 → 扁平列表（前序遍历，保留层级顺序）。
 * 对齐 Java `flattenTree`：命中 keyword 的节点进入结果，但**子节点仍会被遍历**
 * （即父节点被过滤掉时子节点照样可能命中）。
 */
function flattenTree(nodes: TreeNodeVO[], keyword: string | null): BizDataVO[] {
  const result: BizDataVO[] = []
  const walk = (node: TreeNodeVO | null): void => {
    if (node === null) return
    const needle = keyword === null ? null : keyword.toLowerCase()
    const matches =
      needle === null ||
      (node.label !== null && node.label.toLowerCase().includes(needle)) ||
      (node.code !== null && node.code.toLowerCase().includes(needle))
    if (matches) result.push(toDeptRow(node))
    for (const child of node.children ?? []) walk(child)
  }
  for (const node of nodes) walk(node)
  return result
}

/** 部门节点 → BizDataVO。根节点 parentId 为空串（对齐 Java 的三元表达式）。 */
function toDeptRow(node: TreeNodeVO): BizDataVO {
  return bizDataVO(
    String(node.id),
    {
      parentId: node.parentId !== null ? String(node.parentId) : '',
      label: node.label,
      code: node.code,
    },
    null,
    null,
    null,
  )
}

/** 用户 → BizDataVO（字段集合与顺序对齐 Java `toUserRow`）。 */
function toUserRow(user: UserVO): BizDataVO {
  return bizDataVO(
    String(user.id),
    {
      username: user.username,
      nickname: user.nickname,
      orgId: user.orgId,
      orgName: user.orgName,
      status: user.status,
    },
    null,
    null,
    null,
  )
}
